libs/auth-api-lib/src/lib/delegations/delegations-outgoing.service.ts

```diff
@@ -0,0 +1,236 @@
+      ],
+    })
+
+    // TODO: Validate user scope access to the delegation scopes.
+
+    return delegations.map((d) => d.toDTO())
+  }
+
+  async findById(user: User, delegationId: string): Promise<DelegationDTO> {
+    if (!isUuid(delegationId)) {
+      throw new BadRequestException('delegationId must be a valid uuid')
+    }
+
+    const delegation = await this.delegationModel.findOne({
+      where: {
+        id: delegationId,
+        [Op.or]: [
+          { fromNationalId: user.nationalId },
+          { toNationalId: user.nationalId },
+        ],
+      },
+      include: [
+        {
+          model: DelegationScope,
+          as: 'delegationScopes',
+          required: false,
+          where: getScopeValidityWhereClause(DelegationValidity.INCLUDE_FUTURE),
+          include: [
+            {
+              model: ApiScope,
+              as: 'apiScope',
+              where: {
+                enabled: true,
+                allowExplicitDelegationGrant: true,
+              },
+            },
+          ],
+        },
+      ],
+    })
+
+    if (!delegation) {
+      throw new NoContentException()
+    }
+
+    // TODO: Validate user scope access to the delegation scopes.
+
+    return delegation.toDTO()
+  }
+
+  async create(
+    user: User,
+    createDelegation: CreateDelegationDTO,
+  ): Promise<DelegationDTO> {
+    if (createDelegation.toNationalId === user.nationalId) {
+      throw new BadRequestException(`Cannot create delegation to self.`)
+    }
+
+    if (!createDelegation.domainName) {
+      throw new BadRequestException(
+        'Domain name is required to create delegation.',
+      )
+    }
+
+    if (!validateScopesPeriod(createDelegation.scopes)) {
+      throw new BadRequestException(
+        'When scope validTo property is provided it must be in the future',
+      )
+    }
+
+    let delegation = await this.delegationModel.findOne({
+      where: {
+        fromNationalId: user.nationalId,
+        toNationalId: createDelegation.toNationalId,
+        domainName: createDelegation.domainName,
+      },
+    })
+
+    if (!delegation) {
+      delegation = await this.delegationModel.create({
+        id: uuid(),
+        fromNationalId: user.nationalId,
+        toNationalId: createDelegation.toNationalId,
+      })
+    }
+
+    // TODO: User authorization on the scopes
+
+    await this.delegationScopeService.createOrUpdate(
+      delegation.id,
+      createDelegation.scopes,
+    )
+
+    const newDelegation = await this.findById(user, delegation.id)
+
+    if (!newDelegation) {
+      throw new InternalServerErrorException(
+        `Failed to find the newly created delegation with id ${delegation.id}`,
+      )
+    }
+
+    return newDelegation
+  }
+
+  async patch(
+    user: User,
+    delegationId: string,
+    patchedDelegation: PatchDelegationDTO,
+  ): Promise<DelegationDTO> {
+    const currentDelegation = await this.findById(user, delegationId)
+    if (!currentDelegation) {
```

Comment:
Should check `fromNationalId !== user.nationalId`. Also consider the actor?

Reply:
Yes already added the `fromNationalId` check but the actor is a good point, as it would allow the actor to update its own delegation with the subject.